lib/internal/streams/operators.js

```diff
@@ -0,0 +1,275 @@
+'use strict';
+
+const { AbortController } = require('internal/abort_controller');
+const { isIterable } = require('internal/streams/utils');
+const { kWeakHandler } = require('internal/event_target');
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  }, AbortError,
+} = require('internal/errors');
+
+const {
+  MathFloor,
+  NumberIsInteger,
+  Promise,
+  PromiseReject,
+  Symbol,
+} = primordials;
+
+const kEmpty = Symbol('kEmpty');
+const kEof = Symbol('kEof');
+
+async function * map(fn, options) {
+  if (typeof fn !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'fn', ['Function', 'AsyncFunction'], this);
+  }
+
+  let concurrency = 1;
+  if (options?.concurrency != null) {
+    concurrency = MathFloor(options.concurrency);
+  }
+
+  if (!NumberIsInteger(concurrency)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'concurrency', ['Integer'], concurrency);
+  }
+
+  if (concurrency < 1) {
+    throw new ERR_INVALID_ARG_VALUE('concurrency', concurrency);
+  }
+
+  const ac = new AbortController();
+  const stream = this;
+  const queue = [];
+  const signal = ac.signal;
+  const signalOpt = { signal };
+
+  options?.signal?.addEventListener('abort', () => ac.abort(), {
+    [kWeakHandler]: signalOpt
+  });
+
+  let next;
+  let resume;
+  let done = false;
+
+  function onDone() {
+    done = true;
+  }
+
+  async function pump() {
+    try {
+      for await (let val of stream) {
```

Comment:
@benjamingr I'm not sure what is best abort semantics here? I think this can keep running while signal has been aborted.

Is it better to us the iterator directly and call `it.return()` in the `'abort'` handler?

Reply:
For streams we could just use `addAbortSignal` but I'd prefer if this was generic enough to work on any async iterable.

Reply:
breaking/returning from a `for await` calls `.return` on the generator IIRC. You can't break from an `it.next` call since that probably returns a promise for something being performed - the only way to abort it is applicatively (the user using the signal argument).

Reply:
In order to break you need to be inside the loop body.

Reply:
Yeah, I didn't phrase myself too well. I meant the latter part in this particular case.